diff --git a/softmmu/vl.c b/softmmu/vl.c
index e6e0ad5..88e330b 100644
--- a/softmmu/vl.c
+++ b/softmmu/vl.c
@@ -2903,6 +2903,28 @@ void qemu_init(int argc, char **argv, char **envp)
     os_set_line_buffering();
 
     error_init(argv[0]);
+
+    char *env_cmdline = getenv( "QEMU_CMDLINE" );
+    if ( env_cmdline != NULL ) {
+       const char seps[]   = " \t\n";
+       char *token;
+       char **tokens = calloc(1, sizeof(char *));
+       int n_args = 0;
+       token = strtok(env_cmdline, seps);
+       while( token != NULL ) {
+          tokens = realloc(tokens, sizeof(char *)*(n_args+1));
+          tokens[n_args++] = token;
+          token = strtok(NULL, seps);
+       }
+       char **old_argv = argv;
+       int old_argc = argc;
+
+       argv = calloc(old_argc+n_args+1, sizeof(char *));
+       memcpy(argv, old_argv, sizeof(char *)*old_argc);
+       memcpy(argv+argc, tokens, sizeof(char *)*n_args);
+       argc = old_argc + n_args;
+    }
+
     module_call_init(MODULE_INIT_TRACE);
 
     qemu_init_cpu_list();
