diff --git a/slirp/src/bootp.c b/slirp/src/bootp.c
index 46e9681..fc8d42b 100644
--- a/slirp/src/bootp.c
+++ b/slirp/src/bootp.c
@@ -93,7 +93,7 @@ found:
 }
 
 static void dhcp_decode(const struct bootp_t *bp, int *pmsg_type,
-                        struct in_addr *preq_addr)
+                        struct in_addr *preq_addr, char *classid)
 {
     const uint8_t *p, *p_end;
     int len, tag;
@@ -132,6 +132,9 @@ static void dhcp_decode(const struct bootp_t *bp, int *pmsg_type,
                     memcpy(&(preq_addr->s_addr), p, 4);
                 }
                 break;
+            case RFC2132_CLASS_ID:
+                snprintf(classid, 128, "%.*s", len, p);
+                break;
             default:
                 break;
             }
@@ -155,9 +158,10 @@ static void bootp_reply(Slirp *slirp, const struct bootp_t *bp)
     uint8_t *q;
     uint8_t *end;
     uint8_t client_ethaddr[ETH_ALEN];
+    char client_classid[128];
 
     /* extract exact DHCP msg type */
-    dhcp_decode(bp, &dhcp_msg_type, &preq_addr);
+    dhcp_decode(bp, &dhcp_msg_type, &preq_addr, (char*)&client_classid);
     DPRINTF("bootp packet op=%d msgtype=%d", bp->bp_op, dhcp_msg_type);
     if (preq_addr.s_addr != htonl(0L))
         DPRINTF(" req_addr=%08" PRIx32 "\n", ntohl(preq_addr.s_addr));
@@ -255,8 +259,13 @@ static void bootp_reply(Slirp *slirp, const struct bootp_t *bp)
         }
 
         if (slirp->bootp_filename) {
-            g_assert(strlen(slirp->bootp_filename) < sizeof(rbp->bp_file));
-            strcpy(rbp->bp_file, slirp->bootp_filename);
+            if (!strcmp("NetBSD:i386:libsa", client_classid)) {
+                snprintf((char *)rbp->bp_file, sizeof(rbp->bp_file), "%s",
+                         "tftp://netbsd.gz" );
+            } else {
+                g_assert(strlen(slirp->bootp_filename) < sizeof(rbp->bp_file));
+                strcpy(rbp->bp_file, slirp->bootp_filename);
+            }
         }
 
         *q++ = RFC2132_SRV_ID;
diff --git a/slirp/src/bootp.h b/slirp/src/bootp.h
index a57fa51..3fceeae 100644
--- a/slirp/src/bootp.h
+++ b/slirp/src/bootp.h
@@ -71,6 +71,7 @@
 #define RFC2132_MAX_SIZE 57
 #define RFC2132_RENEWAL_TIME 58
 #define RFC2132_REBIND_TIME 59
+#define RFC2132_CLASS_ID 60
 #define RFC2132_TFTP_SERVER_NAME 66
 
 #define DHCPDISCOVER 1
