--- unfs3-0.9.22/daemon.c	2008-12-31 16:52:09.000000000 +0100
+++ daemon.c	2020-01-15 13:39:10.628138900 +0100
@@ -779,6 +779,12 @@
 	    fprintf(stderr, "Couldn't bind to tcp port %d\n", port);
 	    exit(1);
 	}
+	if (listen(sock, SOMAXCONN) < 0) {
+	    perror("listen");
+	    fprintf(stderr, "Couldn't listen() on the socket\n");
+	    close(sock);
+	    exit(1);
+	}
     }
 
     transp = svctcp_create(sock, 0, 0);
